prog/dftb+/lib_dftb/encharges.F90

```diff
@@ -0,0 +1,679 @@
+    real(dp), intent(out), optional :: dqdr(:, :, :)
+
+    !> Derivative of partial charges w.r.t. strain deformations.
+    real(dp), intent(out), optional :: dqdL(:, :, :)
+
+    real(dp), parameter :: small = 1.0e-14_dp
+
+    !> Matrix of 1/R values for each atom pair.
+    real(dp), allocatable :: aMat(:, :)
+    real(dp), allocatable :: xVec(:)
+    real(dp), allocatable :: xFac(:)
+    real(dp), allocatable :: qVec(:)
+
+    real(dp), allocatable :: aInv(:, :)
+    integer, allocatable :: ipiv(:)
+
+    real(dp), allocatable :: aMatdr(:, :, :)
+    real(dp), allocatable :: aMatdL(:, :, :)
+    real(dp), allocatable :: aTrace(:, :)
+
+    logical :: tPeriodic
+    integer :: nDim
+    integer :: iAt1, iSp1, ii, jj
+    real(dp) :: tmp
+
+    tPeriodic = allocated(recPoint)
+
+    nDim = nAtom + 1
+    allocate(xVec(nDim), xFac(nAtom), aMatdr(3, nAtom), aMat(nDim, nDim), aInv(nDim, nDim), &
+        & qVec(nDim), aMatdL(3, 3, nDim), aTrace(3, nAtom), source=0.0_dp)
+    allocate(ipiv(nDim), source=0)
+
+    ! Step 1: contruct RHS for linear system
+    do iAt1 = 1, nAtom
+      iSp1 = species(iAt1)
+      tmp = kcn(iSp1) / (sqrt(cn(iAt1)) + small)
+      xVec(iAt1) = -chi(iSp1) + tmp*cn(iAt1)
+      xFac(iAt1) = -0.5_dp*tmp
+    end do
+    xVec(nDim) = 0.0_dp ! should be charge
+
+    ! Step 2: construct interaction matrix
+    if (tPeriodic) then
+      call getCoulombMatrixPeriodic(nAtom, coords, species, nNeighbour, &
+          & iNeighbour, neighDist2, img2CentCell, recPoint, gam, rad, alpha, &
+          & volume, aMat)
+    else
+      call getCoulombMatrixCluster(nAtom, coords, species, gam, rad, aMat)
+    end if
+    aMat(nDim, 1:nAtom) = 1.0_dp
+    aMat(1:nAtom, nDim) = 1.0_dp
+    aMat(nDim, nDim) = 0.0_dp
+
+    ! Step 3: invert linear system
+    aInv = aMat
+    call sytrf(aInv, ipiv)
+    call sytri(aInv, ipiv)
+    do ii = 1, nDim
+      do jj = ii+1, nDim
+        aInv(ii, jj) = aInv(jj, ii)
+      end do
+    end do
+    call hemv(qVec, aInv, xVec)
+
+    ! Step 4: return atom resolved energies if requested
+    if (present(energies)) then
+      call hemv(xVec, aMat, qVec, alpha=0.5_dp, beta=-1.0_dp)
+      energies = xVec(:nAtom) * qVec(:nAtom)
+    end if
+
+    ! Step 5: get derivative of interaction matrix
+    if (tPeriodic) then
+      call getCoulombDerivsPeriodic(nAtom, coords, species, nNeighbour, iNeighbour, neighDist2,&
+          & img2CentCell, recPoint, alpha, volume, rad, qVec, aMatdr, aMatdL, aTrace)
+    else
+      call getCoulombDerivsCluster(nAtom, coords, species, rad, qVec, aMatdr, aTrace)
+    end if
+    do iAt1 = 1, nAtom
+      aMatdr(:, :, iAt1) = aMatdr(:, :, iAt1) + dcndr(:, :, iAt1) * Xfac(iAt1)
+      aMatdL(:, :, iAt1) = aMatdL(:, :, iAt1) + dcndL(:, :, iAt1) * Xvec(iAt1)
+    end do
+
+    if (present(gradients)) then
+      call gemv(gradients, aMatdr, qVec, beta=1.0_dp)
+    end if
+
+    if (present(stress)) then
+      call gemv(stress, aMatdL, qVec, beta=1.0_dp)
+    end if
```

Comment:
Again, it is not clear, why `beta=1.0_dp` here.

Reply:
Like above.